src/main/java/net/minecraftforge/unification/UnificationRegistry.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.unification;
+
+import javax.annotation.Nullable;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+import com.google.common.base.Preconditions;
+import net.minecraft.util.ResourceLocation;
+import net.minecraftforge.common.ForgeModContainer;
+import net.minecraftforge.oredict.ShapedOreRecipe;
+import net.minecraftforge.oredict.ShapelessOreRecipe;
+
+/**
+ * The {@link UnificationRegistry} allows mods to register *basic* materials so they can be unified.
+ * <p>
+ * Using the {@link UnificationRegistry}, everything can return a single unified type,
+ * so for example players will only ever get one kind of Copper Ingot or Tin Ore.
+ * <p>
+ * Basic materials follow vanilla conventions, like iron ore being smeltable into iron ingots.
+ * If other mods cannot make basic assumptions about your materials, do not register them for unification.
+ * <p>
+ * <b>Usage:</b>
+ * <p>
+ * First, create your {@link UnificationMaterial}s and {@link UnificationForm}s or use the ones listed in {@link UnificationConstants}.
+ * <p>
+ * Then, add your mod's variant with {@link #add(UnificationMaterial, UnificationForm, T)}.
+ * You will get back instances of {@link Unified}, which is used to unify all things with those properties.
+ * <p>
+ * <b>Recipes:</b>
+ * <p>
+ * Recipes created with {@link Unified} for an ingredient will accept any item registered for a type, and return just the one unified type.
+ * {@link Unified} can be used in recipes like {@link ShapedOreRecipe} and {@link ShapelessOreRecipe},
+ * or you can implement your own recipes.
+ * <p>
+ * You can also create JSON recipes that use {@link Unified} as an ingredient or result, by using
+ * the recipe type "forge:ore_shaped" or "forge:ore_shapeless", and using the ingredient format
+ * {"type":"forge:unified", "material": "materialUid", "form": "formUid"}
+ * For examples see the json recipes in the forge test directory "src\test\resources\assets\item_unification_test\recipes"
+ */
+public final class UnificationRegistry<T>
```

Comment:
Should this be a forge registry entry so that they can be looked up with certain keys? (Or kept in registries like that for tile (block) entities)

Reply:
I'm not sure, I think if you have to look up whether one of these registries exists then you probably need to coordinate with the mod that made it pretty closely already. I think mods could expose their unification registry in their API?

Reply:
If this is a forge registry entry, it can be injected with `@ObjectHolder`s too.